pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -422,17 +447,45 @@ func (lrw *logicalReplicationWriterProcessor) checkpoint(
 	return nil
 }
 
+// flushBuffer flushes a buffer using many workers if needed.
+func (lrw *logicalReplicationWriterProcessor) handleStreamBuffer(
+	ctx context.Context, kvs []streampb.StreamEvent_KV,
+) error {
+	unapplied, err := lrw.flushBuffer(ctx, kvs, false)
+	if err != nil {
+		return err
+	}
+	// Put any events that failed to apply into purgatory.
+	if len(unapplied) > 0 {
+		lrw.purgatory.Store(unapplied)
+	}
+
+	return nil
+}
+
+func filterRemaining(kvs []streampb.StreamEvent_KV) []streampb.StreamEvent_KV {
+	failures := kvs
+	var j int
+	for i := range kvs {
+		if len(kvs[i].KeyValue.Key) != 0 {
+			failures[j] = kvs[i]
+			j++
+		}
+	}
+	return failures[:j]
```

Comment:
could crop capacity here too, but i doubt it matters?

Reply:
why would it? we're not passing these to append.